### ISSUE TITLE
Default value of dotprodsimp state variable

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#20239]
During the 1.6 release series performance regressions were identified in some cases involving matrix operations and so the dotprodsimp variable was changed to disable dotprodsimp by default for many operations. This happened in #19588 for master which was then cherry-picked for 1.6.1 in #19612.

Somehow though the two branches ended up with a different default value for the dotprodsimp state variable. On current master we now have:
```julia
In [1]: from sympy.matrices.utilities import _dotprodsimp_state                                                                                

In [2]: _dotprodsimp_state.state                                                                                                               
Out[2]: False
```
Whereas in sympy 1.6.2 we have:
```julia
In [3]: print(_dotprodsimp_state.state)                                                                                                        
None
```

@Pristine-Cat what should be the correct default value for the variable?

I thought that any regressions identified has been fixed in both 1.6.2 and master but running the benchmarks in #20234 I now see that there is a benchmark showing a regression on master (1.7) compared to 1.6.2 which is:
```julia
In [4]: A = Matrix([ 
   ...: [x, 8, 10, 5], 
   ...: [x, 9,  3, 7], 
   ...: [5, 9,  x, 0], 
   ...: [1, 8,  x, 7]])                                                                                                                        

In [5]: %timeit A.det()     # 1.6.2                                                                                                                   
2.75 ms ± 333 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
```julia
In [2]: %timeit A.det()      # 1.7/master                                                                                                                  
16 ms ± 206 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

This diff seems to fix it:
```diff
diff --git a/sympy/matrices/utilities.py b/sympy/matrices/utilities.py
index a65a50b929..cddbe30273 100644
--- a/sympy/matrices/utilities.py
+++ b/sympy/matrices/utilities.py
@@ -9,7 +9,7 @@
 
 class DotProdSimpState(local):
     def __init__(self):
-        self.state = False
+        self.state = None
 
 _dotprodsimp_state = DotProdSimpState()
```
That only gives one failure in matrices (testing the state explicitly) and seems to make some XFAIL tests XPASS:
```console
$ pytest sympy/matrices/ --instafail --tb=short -n4 -d -ra
/Users/enojb/current/sympy/38venv/lib/python3.8/site-packages/_pytest/compat.py:333: PytestDeprecationWarning: The TerminalReporter.writer attribute is deprecated, use TerminalReporter._tw instead at your own risk.
See https://docs.pytest.org/en/latest/deprecations.html#terminalreporter-writer for more information.
  return getattr(object, name, default)
============================================================= test session starts =============================================================
platform darwin -- Python 3.8.1, pytest-5.4.3, py-1.8.1, pluggy-0.13.1
architecture: 64-bit
cache:        yes
ground types: gmpy 2.0.8

rootdir: /Users/enojb/current/sympy/sympy, inifile: pytest.ini
plugins: doctestplus-0.7.0, instafail-0.4.2, xonsh-0.9.13.dev1, cov-2.10.0, profiling-1.7.0, forked-1.1.3, xdist-1.32.0
gw0 [602] / gw1 [602] / gw2 [602] / gw3 [602]
...........................................x.......................................x................................................... [ 22%]
......x........x...................................................x......................................................x............ [ 44%]
.....................................................................................................x.......X.......XX..X............. [ 67%]
...................................................x..........................................x.................................F
______________________________________________________________ test_issue_19809 _______________________________________________________________
[gw1] darwin -- Python 3.8.1 /Users/enojb/current/sympy/38venv/bin/python
sympy/matrices/tests/test_matrices.py:2931: in test_issue_19809
    assert future.result()
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/concurrent/futures/_base.py:432: in result
    return self.__get_result()
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/concurrent/futures/_base.py:388: in __get_result
    raise self._exception
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/concurrent/futures/thread.py:57: in run
    result = self.fn(*self.args, **self.kwargs)
sympy/matrices/tests/test_matrices.py:2923: in f
    assert _dotprodsimp_state.state == False
E   assert None == False
E    +  where None = <sympy.matrices.utilities.DotProdSimpState object at 0x11312e8e0>.state
.....................................................x..............                                                                    [100%]
                                                               DO *NOT* COMMIT!                                                               
=========================================================== short test summary info ===========================================================
XFAIL sympy/matrices/expressions/tests/test_indexing.py::test_block_index_symbolic_fail
XFAIL sympy/matrices/expressions/tests/test_matexpr.py::test_factor_expand
XFAIL sympy/matrices/expressions/tests/test_matmul.py::test_matmul_args_cnc_symbols
XFAIL sympy/matrices/expressions/tests/test_matadd.py::test_matrix_Add_with_scalar
XFAIL sympy/matrices/expressions/tests/test_slice.py::test_symmetry
XFAIL sympy/matrices/tests/test_commonmatrix.py::test_diff
XFAIL sympy/matrices/expressions/tests/test_matexpr.py::test_MatAdd_postprocessor_xfail
XFAIL sympy/matrices/tests/test_eigen.py::test_eigen_vects
XFAIL sympy/matrices/tests/test_matrices.py::test_issue_3979
XFAIL sympy/matrices/tests/test_matrices.py::test_pinv_rank_deficient_when_diagonalization_fails
XPASS sympy/matrices/tests/test_matrices.py::test_issue_17247_expression_blowup_22 
XPASS sympy/matrices/tests/test_matrices.py::test_issue_17247_expression_blowup_8 
XPASS sympy/matrices/tests/test_matrices.py::test_issue_17247_expression_blowup_24 
XPASS sympy/matrices/tests/test_matrices.py::test_issue_17247_expression_blowup_25 
FAILED sympy/matrices/tests/test_matrices.py::test_issue_19809 - assert None == False
======================================= 1 failed, 587 passed, 10 xfailed, 4 xpassed in 93.18s (0:01:33) ====================================
```
Since that makes master consistent with 1.6.2 the last release it would seem safest to make that change so that 1.7 is no different from 1.6.2 but I'm not sure exactly what the intent is of None rather than False.

**Comment by Pristine-Cat** [comment]
`dorprodsimp` False is a hard override and does not allow the optimization on ANY operations and does not allow turning it on via `dotprodsimp=True` in function calls. A value of None is generally off but allows specifying that it be used (such as in some of the tests which are failing due to it not being available).

**Comment by oscarbenjamin** [comment]
Was it supposed to be set to `None`?

I'm not sure how it ended up different on the two branches.

**Comment by Pristine-Cat** [comment]
Yes, otherwise dotprodsimp will never be used even if requested. False was there to allow pre-1.5 behavior in all cases in case of incompatibilities and it can be set to True if you see that it works well in all cases (which at the moment it does not, but maybe in the future if someone fixes it).

**Comment by oscarbenjamin** [comment]
Ok, thanks. I'll set it to `None` then.

The diff also gives a couple of failures in solvers but hopefully not hard to fix:
```
______________________________________________________ test_higher_order_to_first_order _______________________________________________________
[gw1] darwin -- Python 3.8.1 /Users/enojb/current/sympy/38venv/bin/python
sympy/solvers/ode/tests/test_systems.py:1430: in test_higher_order_to_first_order
    assert dsolve(eqs10) == sol10
E   AssertionError: assert [Eq(x(t), C1*...sqrt(47)))/2)] == [Eq(x(t), C1*...(47) + 7))/2)]
E     At index 0 diff: Eq(x(t), C1*(61 - 9*sqrt(47))*sqrt(sqrt(47) + 7)*exp(-t*sqrt(sqrt(47) + 7))/2 + C2*sqrt(7 - sqrt(47))*(61 + 9*sqrt(47))*exp(-t*sqrt(7 - sqrt(47)))/2 - C3*(61 - 9*sqrt(47))*sqrt(sqrt(47) + 7)*exp(t*sqrt(sqrt(47) + 7))/2 - C4*sqrt(7 - sqrt(47))*(61 + 9*sqrt(47))*exp(t*sqrt(7 - sqrt(47)))/2) != Eq(x(t), C1*sqrt(7 - sqrt(47))*(61 + 9*sqrt(47))*exp(-t*sqrt(7 - sqrt(47)))/2 - C2*sqrt(7 - sqrt(47))*(61 + 9*sqrt(47))*exp(t*sqrt(7 - sqrt(47)))/2 + C3*(61 - 9*sqrt(47))*sqrt(sqrt(47) + 7)*exp(-t*sqrt(sqrt(47) + 7))/2 - C4*(61 - 9*sqrt(47))*sqrt(sqrt(47) + 7)*exp(t...
E     
E     ...Full output truncated (2 lines hidden), use '-vv' to show
................x......................x...........................F
______________________________________________________________ test_linodesolve _______________________________________________________________
[gw1] darwin -- Python 3.8.1 /Users/enojb/current/sympy/38venv/bin/python
sympy/solvers/ode/tests/test_systems.py:1819: in test_linodesolve
    assert constant_renumber(linodesolve(A, t, b=b), variables=[t]) == sol
E   AssertionError: assert [-C1*exp(-t/2...5)*t/2)/5, t)] == [-C1*exp(-t/2...5)*t/2)/5, t)]
E     At index 0 diff: -C1*exp(-t/2 + sqrt(5)*t/2)/2 + sqrt(5)*C1*exp(-t/2 + sqrt(5)*t/2)/2 - sqrt(5)*C2*exp(-sqrt(5)*t/2 - t/2)/2 - C2*exp(-sqrt(5)*t/2 - t/2)/2 - exp(-t/2 + sqrt(5)*t/2)*Integral(t*exp(-sqrt(5)*t/2 + t/2)/(-5 + sqrt(5)) - sqrt(5)*t*exp(-sqrt(5)*t/2 + t/2)/(-5 + sqrt(5)), t)/2 + sqrt(5)*exp(-t/2 + sqrt(5)*t/2)*Integral(t*exp(-sqrt(5)*t/2 + t/2)/(-5 + sqrt(5)) - sqrt(5)*t*exp(-sqrt(5)*t/2 + t/2)/(-5 + sqrt(5)), t)/2 - sqrt(5)*exp(-sqrt(5)*t/2 - t/2)*Integral(-sqrt(5)*t*exp(t/2 + sqrt(5)*t/2)/5, t)/2 - exp(-sqrt(5)*t/2 - t/2)*Integral(-sqrt(5)*t*exp(t/2 + sqrt(...
E     
E     ...Full output truncated (2 lines hidden), use '-vv' to show
```

**Comment by Pristine-Cat** [comment]
My guess is someone added those and could only get them running with `dotprodsimp` set to `False`, which is why they changed it globally. Try wrapping the failing ops with the dotprodsimp context manager with `False`.

**Comment by oscarbenjamin** [comment]
For the det case the slowdown I investigate is I think that when dotprodsimp is off then cancel is used:
https://github.com/sympy/sympy/blob/2513aca3af9d5c85671f155dfe5de01ccd8b77df/sympy/matrices/determinant.py#L657-L661
Since `cancel` is also the part of `dotprodsimp` that can be slow I expect that both cases would suffer for some inputs.

**Comment by oscarbenjamin** [comment]
> My guess is

If you look at the diff in #19588 you can see that it is set to False whereas in #19612 it is set to None. Since I created #19612 using `git cherry-pick` I'm really not sure how that happened.

**Comment by Pristine-Cat** [comment]
> Since `cancel` is also the part of `dotprodsimp` that can be slow I expect that both cases would suffer for some inputs.

A pure `cancel` in theory can be slower than `dotprodsimp` since the latter can reduce the expression a bit before passing it on to the `cancel`, so I don't think it makes sense to do `cancel` alone.

**Comment by Pristine-Cat** [comment]
> https://github.com/sympy/sympy/blob/2513aca3af9d5c85671f155dfe5de01ccd8b77df/sympy/matrices/determinant.py#L657-L661

Actually for this case what is happening is that if `dotprodsimp` was set `False` globally then the `_get_intermediate_simp_bool` would never proc and it would always fall through to the `cancel`. With global `None` it should always proc due to the `True` arg and that second part should with `cancel` in that case should not be necessary.

**Comment by Pristine-Cat** [comment]
Actually never mind, I think I am the one responsible for setting it `False` globally when those compatibility problems popped up to revert to pre-1.5, but `None` is better as it allows optimizations in a few cases like that bareiss. Sorry...

**Comment by oscarbenjamin** [comment]
No problem. This seems easy enough to change. Could you review #20240?

**Comment by oscarbenjamin** [comment]
The benchmarks have been pretty useful for spotting unintentional changes

**Comment by oscarbenjamin** [comment]
I've merged that change for 1.7 and master. Thanks for the help @Pristine-Cat

**Comment by Pristine-Cat** [comment]
No problem. One thing, I remember now is the change to `False` was because of #19532. I mention this because with `None` that and similar will probably pop up again. In which case, just tell people to wrap code in context manager with `False`.

**Comment by oscarbenjamin** [comment]
It was because of #19532 that we ended up making the changes to fix the issue in 1.6.1 and also on master but we somehow ended up with a different version of the fix on each branch.

It looks like that slowdown didn't get fixed in 1.6.2 but the option to not use dotprodsimp meant that the issue was sort of fixed. Then on master we had dotprodsimp fully disabled by default so actually that matrix could run in a reasonable time. It seems now that #20240 has made that particular example slow again. There is now the option though to disable dotprodsimp.

The safest thing for now I think is to go with what was released in 1.6.2 but I guess that means that #19887 needs to be reopened.
